### PR TITLE
feat: deposit

### PR DESCRIPTION
**Pull request by TheAustrian1998**
- re-order contracts and inheritance
- add code for deposit()
- tests

### CODE REVIEW

File: contracts/DCAStrategies/DCAStrategies/DCAStrategies.sol [hunk 1]
```diff
@@ -4,9 +4,8 @@ pragma solidity >=0.8.7 <0.9.0;
 import './DCAStrategiesManagementHandler.sol';
 import './DCAStrategiesPermissionsHandler.sol';
 import './DCAStrategiesPositionsHandler.sol';
-import '../../interfaces/IDCAStrategies.sol';
```
[the remark this answers is not in the document]
Comment (TheAustrian1998) [reply contracts/DCAStrategies/DCAStrategies/DCAStrategies.sol:7]: Yes, captain!

File: contracts/mocks/DCAStrategies/DCAStrategiesPositionsHandlerMock.sol [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      _tokenShares[_tokenShares.length - 1].token = _tokens[i].token;
+      _tokenShares[_tokenShares.length - 1].share = _tokens[i].share;
+    }
+  }
+
+  function _getTokenShares(uint80 _strategyId, uint16 _version) internal view override returns (IDCAStrategies.ShareOfToken[] memory) {
+    _strategyId; // shh
+    _version; // shh
+    return _tokenShares;
+  }
```
Comment (nchamo) [inline contracts/mocks/DCAStrategies/DCAStrategiesPositionsHandlerMock.sol:31]: Haha, loved the shh. You can avoid that like this:
```suggestion
  function _getTokenShares(uint80, uint16) internal view override returns (IDCAStrategies.ShareOfToken[] memory) {
    return _tokenShares;
  }
```

File: contracts/mocks/DCAStrategies/DCAStrategiesPositionsHandlerMock.sol [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  function getCreateCalls() external view returns (CreateCall[] memory) {
+    return _createCalls;
+  }
+
+  function setTokenShares(IDCAStrategies.ShareOfToken[] calldata _tokens) external {
+    _tokenShares.push();
+    for (uint256 i = 0; i < _tokens.length; i++) {
+      _tokenShares[_tokenShares.length - 1].token = _tokens[i].token;
+      _tokenShares[_tokenShares.length - 1].share = _tokens[i].share;
+    }
```
Comment (nchamo) [inline contracts/mocks/DCAStrategies/DCAStrategiesPositionsHandlerMock.sol:24]: When you have an array of a struct that inside has an array, you need to do this. You need to push something to the first position, and then fill the array inside the struct (like in `_create`). But here, you just have a normal array
```suggestion
    for (uint256 i = 0; i < _tokens.length; i++) {
      _tokenShares.push(_tokens[i]);
    }
```

File: test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    SHARE_TOKEN_A = { token: tokenA.address, share: BigNumber.from(50e2) };
+    SHARE_TOKEN_B = { token: tokenB.address, share: BigNumber.from(50e2) };
+    SHARES = [SHARE_TOKEN_A, SHARE_TOKEN_B];
+  });
+
+  beforeEach(async () => {
+    await snapshot.revert(snapshotId);
+  });
+
+  describe('deposit', () => {
```
Comment (nchamo) [inline test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts:38]: Could you add a test that makes sure that transferFrom was called correctly? Like [here](https://github.com/Mean-Finance/swappers/blob/main/test/unit/swap-adapter.spec.ts#L88)

File: contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol [hunk 5]
```diff
@@ -7,7 +7,45 @@ abstract contract DCAStrategiesPositionsHandler is IDCAStrategiesPositionsHandle
   // TODO: add function similar to this one https://github.com/Mean-Finance/dca-v2-core/blob/main/contracts/interfaces/IDCAHub.sol#L243
 
   /// @inheritdoc IDCAStrategiesPositionsHandler
-  function deposit(IDCAStrategies.DepositParams calldata parameters) external override returns (uint256) {}
+  function deposit(IDCAStrategies.DepositParams calldata parameters) external returns (uint256) {
+    IDCAStrategies.ShareOfToken[] memory _tokens = _getTokenShares(parameters.strategyId, parameters.version);
+    if (_tokens.length == 0) revert InvalidStrategy();
+
+    IERC20(parameters.from).transferFrom(msg.sender, address(this), parameters.amount);
```
Comment (nchamo) [inline contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol:14]: Let's use `safeTransferFrom` here, like [this](https://github.com/Mean-Finance/swappers/blob/main/solidity/contracts/SwapAdapter.sol#L31)

File: test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    let tx: TransactionResponse;
+    let toDeposit: BigNumber = ethers.utils.parseUnits('300');
+    let amountOfSwaps: number = 5;
+    let swapInterval: number = 7 * 24 * 60 * 60; // 1 week
```
Comment (nchamo) [inline test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts:42]: Just as an fyi, when you assign a value, typescript can infer the type
```suggestion
    let amountOfSwaps = 5;
    let swapInterval = 7 * 24 * 60 * 60; // 1 week
```

**Review by nchamo — COMMENTED** [review]
Great work! Just a few minor changes more and we should be good to go

File: contracts/mocks/DCAStrategies/DCAStrategiesPositionsHandlerMock.sol (hunk 2 above)
Comment (TheAustrian1998) [reply contracts/mocks/DCAStrategies/DCAStrategiesPositionsHandlerMock.sol:31]: Cool trick

**Review by nchamo — APPROVED** [review]
Great work!

**Pull request merged**